src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2680,21 +2690,73 @@ class ShenandoahUpdateHeapRefsTask : public AbstractGangTask {
+        // an old-gen region as part of an old-gen collection.  Otherwise, a subseqent update-refs scan of
+        // the same region will see stale pointers and crash.
+        //
+        //   r->top() represents the upper end of memory that has been allocated within this region.
+        //       As new objects are allocated, the value of r->top() increases to accomodate each new
+        //       object.
+        //   At the start of evacuation, "update_watermark" is initalized to represent the value of top().
+        //       Objects newly allocated during evacuation do not need to be visited during update-refs
+        //       because the to-space invariant which is in force throughout evacuation assures that no from-space
+        //       pointer is written to any newly allocated object.  In the case that survivor objects are evacuated
+        //       into this region during evacuation, the region's watermark is incremented to represent the end of
+        //       of the memory range known to hold newly evacuated objects.  Note that incrementing watermark to
+        //       account for objects newly evacuated into the region may result in otherwise unnecessary updating
+        //       of references contained within newly allocated objects that happen to be located between the
+        //       initial value of watermark and the updated value of watermark.
+
+        HeapWord *p = r->bottom();
+        ShenandoahObjectToOopBoundedClosure<T> objs(&cl, p, update_watermark);
+
+        // TODO: This code assumes every object ever allocated within this old-gen region is still live.  If we
+        // allow a sweep phase to turn garbage objects into free memory regions, we need to modify this code to
+        // skip over and/or synchronize access to these free memory regions.  There might be races, for example,
+        // if we are trying to scan one of these free memory regions while a different thread is trying to
+        // allocate from within a free region.
+        //
+        // Alternative approaches are also under consideration.  For example:
+        //  1. Coalesce, fill, and register each range of contiguous dead objects so that subsequent updating of
+        //     references can be done more efficiently.
+        //  2. Retain the mark bitmap from the most recently completed old GC effort and use this bitmap to allow
+        //     skipping over objects that were not live as of the most recently completed old-gen GC effort.
+
+        // Anything beyond update_watermark does not need to be updated.
+        while (p < update_watermark) {
+          oop obj = oop(p);
+
+          // The invocation of do_object() is "borrowed" from the implementation of
+          // ShenandoahHeap::marked_object_iterate(), which is called by _heap->marked_object_oop_iterate().
+          objs.do_object(obj);
+          p += obj->size();
```

Comment:
As noted before, this might crash, if the Klass* that an unreachable object points to has been unloaded. The problem, as far as I can tell, arises because we don't have up-to-date marking information anymore, because we clean the whole bitmap, and subsequent young-gen collection doesn't establish old-gen reachability. Could we solve this by preserving the most-recent old-gen marking in the bitmap, and only clean the young-gen-parts? For now, it can probably be worked around by turning off ClassUnloading altogether.

Reply:
We've experimented with a preserved bitmap on a branch. It works until we hit a full collection, at which point things go sideways. The compaction phase specifically needed to be modified to maintain the correctness of the bitmap we wanted to preserve. It needs a bit more debugging, but it's been tabled for now.